File: cathode-sync/src/main/java/net/simonvt/cathode/sync/scheduler/EpisodeTaskScheduler.java
/*
 * Copyright (C) 2017 Simon Vig Therkildsen
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.simonvt.cathode.sync.scheduler;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.text.format.DateUtils;
import java.io.IOException;
import javax.inject.Inject;
import javax.inject.Singleton;
import net.simonvt.cathode.api.body.SyncItems;
import net.simonvt.cathode.api.service.CheckinService;
import net.simonvt.cathode.api.util.TimeUtils;
import net.simonvt.cathode.common.database.Cursors;
import net.simonvt.cathode.common.event.ErrorEvent;
import net.simonvt.cathode.jobqueue.JobManager;
import net.simonvt.cathode.provider.DatabaseContract.EpisodeColumns;
import net.simonvt.cathode.provider.DatabaseContract.ShowColumns;
import net.simonvt.cathode.provider.ProviderSchematic.Episodes;
import net.simonvt.cathode.provider.ProviderSchematic.Shows;
import net.simonvt.cathode.provider.helper.EpisodeDatabaseHelper;
import net.simonvt.cathode.provider.helper.ShowDatabaseHelper;
import net.simonvt.cathode.provider.util.DataHelper;
import net.simonvt.cathode.remote.action.RemoveHistoryItem;
import net.simonvt.cathode.remote.action.shows.AddEpisodeToHistory;
import net.simonvt.cathode.remote.action.shows.CollectEpisode;
import net.simonvt.cathode.remote.action.shows.RateEpisode;
import net.simonvt.cathode.remote.action.shows.RemoveEpisodeFromHistory;
import net.simonvt.cathode.remote.action.shows.WatchlistEpisode;
import net.simonvt.cathode.remote.sync.SyncWatching;
import net.simonvt.cathode.settings.TraktLinkSettings;
import net.simonvt.cathode.sync.R;
import net.simonvt.cathode.sync.trakt.CheckIn;
import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Response;
import timber.log.Timber;

@Singleton public class EpisodeTaskScheduler extends BaseTaskScheduler {

  private ShowDatabaseHelper showHelper;
  private EpisodeDatabaseHelper episodeHelper;
  private CheckinService checkinService;
  private CheckIn checkIn;

  @Inject
  public EpisodeTaskScheduler(Context context, JobManager jobManager, ShowDatabaseHelper showHelper,
      EpisodeDatabaseHelper episodeHelper, CheckinService checkinService, CheckIn checkIn) {
    super(context, jobManager);
    this.showHelper = showHelper;
    this.episodeHelper = episodeHelper;
    this.checkinService = checkinService;
    this.checkIn = checkIn;
  }

  public void addToHistoryNow(final long episodeId) {
    addToHistory(episodeId, System.currentTimeMillis());
  }

  private Cursor getOlderEpisodes(long episodeId) {
    long showId = episodeHelper.getShowId(episodeId);
    int season = episodeHelper.getSeason(episodeId);
    int episode = episodeHelper.getNumber(episodeId);
    Cursor episodes = context.getContentResolver().query(Episodes.EPISODES, new String[] {
        EpisodeColumns.ID, EpisodeColumns.SEASON, EpisodeColumns.EPISODE,
    }, EpisodeColumns.SHOW_ID
        + "="
        + showId
        + " AND "
        + EpisodeColumns.WATCHED
        + "=0 AND "
        + EpisodeColumns.SEASON
        + ">0 AND (("
        + EpisodeColumns.SEASON
        + "="
        + season
        + " AND "
        + EpisodeColumns.EPISODE
        + "<"
        + episode
        + ") OR "
        + EpisodeColumns.SEASON
        + "<"
        + season
        + ")", null, null);

    return episodes;
  }

  public void addOlderToHistoryNow(final long episodeId) {
    execute(new Runnable() {
      @Override public void run() {
        Cursor episodes = getOlderEpisodes(episodeId);
        while (episodes.moveToNext()) {
          long id = Cursors.getLong(episodes, EpisodeColumns.ID);
          addToHistoryNow(id);
        }
        episodes.close();
      }
    });
  }

  public void addToHistoryOnRelease(final long episodeId) {
    addToHistory(episodeId, SyncItems.TIME_RELEASED);
  }

  public void addOlderToHistoryOnRelease(final long episodeId) {
    execute(new Runnable() {
      @Override public void run() {
        Cursor episodes = getOlderEpisodes(episodeId);
        while (episodes.moveToNext()) {
          long id = Cursors.getLong(episodes, EpisodeColumns.ID);
          addToHistory(id, SyncItems.TIME_RELEASED);
        }
        episodes.close();
      }
    });
  }

  public void addOlderToHistory(final long episodeId, final int year, final int month,
      final int day, final int hour, final int minute) {
    execute(new Runnable() {
      @Override public void run() {
        Cursor episodes = getOlderEpisodes(episodeId);
        while (episodes.moveToNext()) {
          long id = Cursors.getLong(episodes, EpisodeColumns.ID);
          addToHistory(id, TimeUtils.getMillis(year, month, day, hour, minute));
        }
        episodes.close();
      }
    });
  }

  public void addToHistory(final long episodeId, final long watchedAt) {
    final String isoWhen = TimeUtils.getIsoTime(watchedAt);
    addToHistory(episodeId, isoWhen);
  }

  public void addToHistory(final long episodeId, final int year, final int month, final int day,
      final int hour, final int minute) {
    addToHistory(episodeId, TimeUtils.getMillis(year, month, day, hour, minute));
  }

  public void addToHistory(final long episodeId, final String watchedAt) {
    execute(new Runnable() {
      @Override public void run() {
        Cursor c = episodeHelper.query(episodeId, EpisodeColumns.SHOW_ID, EpisodeColumns.SEASON,
            EpisodeColumns.EPISODE);
        c.moveToFirst();
        final long showId = Cursors.getLong(c, EpisodeColumns.SHOW_ID);
        final long traktId = showHelper.getTraktId(showId);
        final int season = Cursors.getInt(c, EpisodeColumns.SEASON);
        final int number = Cursors.getInt(c, EpisodeColumns.EPISODE);
        c.close();

        if (SyncItems.TIME_RELEASED.equals(watchedAt)) {
          episodeHelper.addToHistory(episodeId, EpisodeDatabaseHelper.WATCHED_RELEASE);
        } else {
          episodeHelper.addToHistory(episodeId, TimeUtils.getMillis(watchedAt));
        }

        if (TraktLinkSettings.isLinked(context)) {
          queue(new AddEpisodeToHistory(traktId, season, number, watchedAt));
        }
      }
    });
  }

  public void removeFromHistory(final long episodeId) {
    execute(new Runnable() {
      @Override public void run() {
        episodeHelper.removeFromHistory(episodeId);

        if (TraktLinkSettings.isLinked(context)) {
          Cursor c = episodeHelper.query(episodeId, EpisodeColumns.SHOW_ID, EpisodeColumns.SEASON,
              EpisodeColumns.EPISODE);
          c.moveToFirst();
          final long showId = Cursors.getLong(c, EpisodeColumns.SHOW_ID);
          final long traktId = showHelper.getTraktId(showId);
          final int season = Cursors.getInt(c, EpisodeColumns.SEASON);
          final int number = Cursors.getInt(c, EpisodeColumns.EPISODE);
          c.close();

          queue(new RemoveEpisodeFromHistory(traktId, season, number));
        }
      }
    });
  }

  public void removeHistoryItem(final long episodeId, final long historyId,
      final boolean lastItem) {
    execute(new Runnable() {
      @Override public void run() {
        if (lastItem) {
          episodeHelper.removeFromHistory(episodeId);
        }

        if (TraktLinkSettings.isLinked(context)) {
          queue(new RemoveHistoryItem(historyId));
        }
      }
    });
  }

  public void checkin(final long episodeId, final String message, final boolean facebook,
      final boolean twitter, final boolean tumblr) {
    execute(new Runnable() {
      @Override public void run() {
        Cursor watching =
            context.getContentResolver().query(Episodes.EPISODE_WATCHING, new String[] {
                EpisodeColumns.ID, EpisodeColumns.EXPIRES_AT,
            }, null, null, null);

        final long currentTime = System.currentTimeMillis();
        long expires = 0;
        if (watching.moveToFirst()) {
          expires = Cursors.getLong(watching, EpisodeColumns.EXPIRES_AT);
        }

        Cursor episode =
            episodeHelper.query(episodeId, EpisodeColumns.SHOW_ID, EpisodeColumns.TITLE,
                EpisodeColumns.SEASON, EpisodeColumns.EPISODE, EpisodeColumns.WATCHED);
        episode.moveToFirst();
        final long showId = Cursors.getLong(episode, EpisodeColumns.SHOW_ID);
        final int season = Cursors.getInt(episode, EpisodeColumns.SEASON);
        final int number = Cursors.getInt(episode, EpisodeColumns.EPISODE);
        final boolean watched = Cursors.getBoolean(episode, EpisodeColumns.WATCHED);
        final String title = DataHelper.getEpisodeTitle(context, episode, season, number, watched);
        episode.close();

        Cursor show = context.getContentResolver().query(Shows.withId(showId), new String[] {
            ShowColumns.RUNTIME,
        }, null, null, null);
        show.moveToFirst();
        final int runtime = Cursors.getInt(show, ShowColumns.RUNTIME);
        final long watchSlop = (long) (runtime * DateUtils.MINUTE_IN_MILLIS * 0.8f);
        show.close();

        if (watching.getCount() == 0 || ((expires - watchSlop) < currentTime && expires > 0)) {
          if (checkIn.episode(episodeId, message, facebook, twitter, tumblr)) {
            episodeHelper.checkIn(episodeId);
          }
        } else {
          ErrorEvent.post(context.getString(R.string.checkin_error_watching, title));
        }

        watching.close();
        queue(new SyncWatching());
      }
    });
  }

  public void cancelCheckin() {
    execute(new Runnable() {
      @Override public void run() {
        Cursor episode = null;
        try {
          episode = context.getContentResolver().query(Episodes.EPISODE_WATCHING, new String[] {
              EpisodeColumns.ID, EpisodeColumns.STARTED_AT, EpisodeColumns.EXPIRES_AT,
          }, null, null, null);

          if (episode.moveToFirst()) {
            final long id = Cursors.getLong(episode, EpisodeColumns.ID);
            final long startedAt = Cursors.getLong(episode, EpisodeColumns.STARTED_AT);
            final long expiresAt = Cursors.getLong(episode, EpisodeColumns.EXPIRES_AT);

            ContentValues values = new ContentValues();
            values.put(EpisodeColumns.CHECKED_IN, false);
            context.getContentResolver().update(Episodes.EPISODE_WATCHING, values, null, null);

            try {
              Call<ResponseBody> call = checkinService.deleteCheckin();
              Response<ResponseBody> response = call.execute();
              if (response.isSuccessful()) {
                return;
              }
            } catch (IOException e) {
              Timber.d(e);
            }

            ErrorEvent.post(context.getString(R.string.checkin_cancel_error));

            values.clear();
            values.put(EpisodeColumns.CHECKED_IN, true);
            values.put(EpisodeColumns.STARTED_AT, startedAt);
            values.put(EpisodeColumns.EXPIRES_AT, expiresAt);
            context.getContentResolver().update(Episodes.withId(id), values, null, null);

            queue(new SyncWatching());
          }
        } finally {
          if (episode != null) {
            episode.close();
          }
        }
      }
    });
  }

  /**
   * Add episodes to user library collection.
   *
   * @param episodeId The database id of the episode.
   */
  public void setIsInCollection(final long episodeId, final boolean inCollection) {
    execute(new Runnable() {
      @Override public void run() {
        String collectedAt = null;
        long collectedAtMillis = 0L;
        if (inCollection) {
          collectedAt = TimeUtils.getIsoTime();
          collectedAtMillis = TimeUtils.getMillis(collectedAt);
        }

        episodeHelper.setInCollection(episodeId, inCollection, collectedAtMillis);

        if (TraktLinkSettings.isLinked(context)) {
          Cursor c = episodeHelper.query(episodeId, EpisodeColumns.SHOW_ID, EpisodeColumns.SEASON,
              EpisodeColumns.EPISODE);
          c.moveToFirst();
          final long showId = Cursors.getLong(c, EpisodeColumns.SHOW_ID);
          final long traktId = showHelper.getTraktId(showId);
          final int season = Cursors.getInt(c, EpisodeColumns.SEASON);
          final int number = Cursors.getInt(c, EpisodeColumns.EPISODE);
          c.close();

          queue(new CollectEpisode(traktId, season, number, inCollection, collectedAt));
        }
      }
    });
  }

  public void setIsInWatchlist(final long episodeId, final boolean inWatchlist) {
    execute(new Runnable() {
      @Override public void run() {
        String listedAt = null;
        long listeddAtMillis = 0L;
        if (inWatchlist) {
          listedAt = TimeUtils.getIsoTime();
          listeddAtMillis = TimeUtils.getMillis(listedAt);
        }

        episodeHelper.setIsInWatchlist(episodeId, inWatchlist, listeddAtMillis);

        if (TraktLinkSettings.isLinked(context)) {
          Cursor c = episodeHelper.query(episodeId, EpisodeColumns.SHOW_ID, EpisodeColumns.SEASON,
              EpisodeColumns.EPISODE);
          c.moveToFirst();
          final long showId = Cursors.getLong(c, EpisodeColumns.SHOW_ID);
          final long traktId = showHelper.getTraktId(showId);
          final int season = Cursors.getInt(c, EpisodeColumns.SEASON);
          final int number = Cursors.getInt(c, EpisodeColumns.EPISODE);
          c.close();

          queue(new WatchlistEpisode(traktId, season, number, inWatchlist, listedAt));
        }
      }
    });
  }

  /**
   * Rate an episode on trakt. Depending on the user settings, this will also send out social
   * updates to facebook,
   * twitter, and tumblr.
   *
   * @param episodeId The database id of the episode.
   * @param rating A rating betweeo 1 and 10. Use 0 to undo rating.
   */
  public void rate(final long episodeId, final int rating) {
    execute(new Runnable() {
      @Override public void run() {
        String ratedAt = TimeUtils.getIsoTime();
        long ratedAtMillis = TimeUtils.getMillis(ratedAt);

        final long showId = episodeHelper.getShowId(episodeId);
        final long showTraktId = showHelper.getTraktId(showId);
        Cursor c = episodeHelper.query(episodeId, EpisodeColumns.EPISODE, EpisodeColumns.SEASON);

        if (c.moveToFirst()) {
          final int episode = Cursors.getInt(c, EpisodeColumns.EPISODE);
          final int season = Cursors.getInt(c, EpisodeColumns.SEASON);

          ContentValues values = new ContentValues();
          values.put(EpisodeColumns.USER_RATING, rating);
          values.put(EpisodeColumns.RATED_AT, ratedAtMillis);
          context.getContentResolver().update(Episodes.withId(episodeId), values, null, null);

          if (TraktLinkSettings.isLinked(context)) {
            queue(new RateEpisode(showTraktId, season, episode, rating, ratedAt));
          }
        }
        c.close();
      }
    });
  }
}

File: cathode-sync/src/main/java/net/simonvt/cathode/sync/scheduler/EpisodeTaskScheduler.kt
/*
 * Copyright (C) 2017 Simon Vig Therkildsen
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.simonvt.cathode.sync.scheduler

import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.text.format.DateUtils
import androidx.work.WorkManager
import kotlinx.coroutines.launch
import net.simonvt.cathode.api.body.SyncItems
import net.simonvt.cathode.api.service.CheckinService
import net.simonvt.cathode.api.util.TimeUtils
import net.simonvt.cathode.common.database.getBoolean
import net.simonvt.cathode.common.database.getInt
import net.simonvt.cathode.common.database.getLong
import net.simonvt.cathode.common.event.ErrorEvent
import net.simonvt.cathode.jobqueue.JobManager
import net.simonvt.cathode.provider.DatabaseContract.EpisodeColumns
import net.simonvt.cathode.provider.DatabaseContract.ShowColumns
import net.simonvt.cathode.provider.ProviderSchematic.Episodes
import net.simonvt.cathode.provider.ProviderSchematic.Shows
import net.simonvt.cathode.provider.helper.EpisodeDatabaseHelper
import net.simonvt.cathode.provider.helper.ShowDatabaseHelper
import net.simonvt.cathode.provider.query
import net.simonvt.cathode.provider.util.DataHelper
import net.simonvt.cathode.remote.action.RemoveHistoryItem
import net.simonvt.cathode.remote.action.shows.AddEpisodeToHistory
import net.simonvt.cathode.remote.action.shows.CollectEpisode
import net.simonvt.cathode.remote.action.shows.RateEpisode
import net.simonvt.cathode.remote.action.shows.RemoveEpisodeFromHistory
import net.simonvt.cathode.remote.action.shows.WatchlistEpisode
import net.simonvt.cathode.settings.TraktLinkSettings
import net.simonvt.cathode.sync.R
import net.simonvt.cathode.sync.trakt.CheckIn
import net.simonvt.cathode.work.enqueueNow
import net.simonvt.cathode.work.user.SyncWatchingWorker
import timber.log.Timber
import java.io.IOException
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class EpisodeTaskScheduler @Inject constructor(
  context: Context,
  jobManager: JobManager,
  private val workManager: WorkManager,
  private val showHelper: ShowDatabaseHelper,
  private val episodeHelper: EpisodeDatabaseHelper,
  private val checkinService: CheckinService,
  private val checkIn: CheckIn
) : BaseTaskScheduler(context, jobManager) {

  fun addToHistoryNow(episodeId: Long) {
    addToHistory(episodeId, System.currentTimeMillis())
  }

  private fun getOlderEpisodes(episodeId: Long): Cursor {
    val showId = episodeHelper.getShowId(episodeId)
    val season = episodeHelper.getSeason(episodeId)
    val episode = episodeHelper.getNumber(episodeId)

    return context.contentResolver.query(
      Episodes.EPISODES,
      arrayOf(EpisodeColumns.ID, EpisodeColumns.SEASON, EpisodeColumns.EPISODE),
      EpisodeColumns.SHOW_ID + "=" + showId + " AND " +
          EpisodeColumns.WATCHED + "=0 AND " +
          EpisodeColumns.SEASON + ">0 AND ((" +
          EpisodeColumns.SEASON + "=" + season + " AND " +
          EpisodeColumns.EPISODE + "<" + episode + ") OR " +
          EpisodeColumns.SEASON + "<" + season + ")"
    )
  }

  fun addOlderToHistoryNow(episodeId: Long) {
    scope.launch {
      val episodes = getOlderEpisodes(episodeId)
      while (episodes.moveToNext()) {
        val id = episodes.getLong(EpisodeColumns.ID)
        addToHistoryNow(id)
      }
      episodes.close()
    }
  }

  fun addToHistoryOnRelease(episodeId: Long) {
    addToHistory(episodeId, SyncItems.TIME_RELEASED)
  }

  fun addOlderToHistoryOnRelease(episodeId: Long) {
    scope.launch {
      val episodes = getOlderEpisodes(episodeId)
      while (episodes.moveToNext()) {
        val id = episodes.getLong(EpisodeColumns.ID)
        addToHistory(id, SyncItems.TIME_RELEASED)
      }
      episodes.close()
    }
  }

  fun addOlderToHistory(
    episodeId: Long,
    year: Int,
    month: Int,
    day: Int,
    hour: Int,
    minute: Int
  ) {
    scope.launch {
      val episodes = getOlderEpisodes(episodeId)
      while (episodes.moveToNext()) {
        val id = episodes.getLong(EpisodeColumns.ID)
        addToHistory(id, TimeUtils.getMillis(year, month, day, hour, minute))
      }
      episodes.close()
    }
  }

  fun addToHistory(episodeId: Long, watchedAt: Long) {
    val isoWhen = TimeUtils.getIsoTime(watchedAt)
    addToHistory(episodeId, isoWhen)
  }

  fun addToHistory(episodeId: Long, year: Int, month: Int, day: Int, hour: Int, minute: Int) {
    addToHistory(episodeId, TimeUtils.getMillis(year, month, day, hour, minute))
  }

  fun addToHistory(episodeId: Long, watchedAt: String) {
    scope.launch {
      val c = context.contentResolver.query(
        Episodes.withId(episodeId),
        arrayOf(EpisodeColumns.SHOW_ID, EpisodeColumns.SEASON, EpisodeColumns.EPISODE)
      )
      c.moveToFirst()
      val showId = c.getLong(EpisodeColumns.SHOW_ID)
      val traktId = showHelper.getTraktId(showId)
      val season = c.getInt(EpisodeColumns.SEASON)
      val number = c.getInt(EpisodeColumns.EPISODE)
      c.close()

      if (SyncItems.TIME_RELEASED == watchedAt) {
        episodeHelper.addToHistory(episodeId, EpisodeDatabaseHelper.WATCHED_RELEASE)
      } else {
        episodeHelper.addToHistory(episodeId, TimeUtils.getMillis(watchedAt))
      }

      if (TraktLinkSettings.isLinked(context)) {
        queue(AddEpisodeToHistory(traktId, season, number, watchedAt))
      }
    }
  }

  fun removeFromHistory(episodeId: Long) {
    scope.launch {
      episodeHelper.removeFromHistory(episodeId)

      if (TraktLinkSettings.isLinked(context)) {
        val c = context.contentResolver.query(
          Episodes.withId(episodeId),
          arrayOf(EpisodeColumns.SHOW_ID, EpisodeColumns.SEASON, EpisodeColumns.EPISODE)
        )
        c.moveToFirst()
        val showId = c.getLong(EpisodeColumns.SHOW_ID)
        val traktId = showHelper.getTraktId(showId)
        val season = c.getInt(EpisodeColumns.SEASON)
        val number = c.getInt(EpisodeColumns.EPISODE)
        c.close()

        queue(RemoveEpisodeFromHistory(traktId, season, number))
      }
    }
  }

  fun removeHistoryItem(episodeId: Long, historyId: Long, lastItem: Boolean) {
    scope.launch {
      if (lastItem) {
        episodeHelper.removeFromHistory(episodeId)
      }

      if (TraktLinkSettings.isLinked(context)) {
        queue(RemoveHistoryItem(historyId))
      }
    }
  }

  fun checkin(
    episodeId: Long,
    message: String?,
    facebook: Boolean,
    twitter: Boolean,
    tumblr: Boolean
  ) {
    scope.launch {
      val watching = context.contentResolver.query(
        Episodes.EPISODE_WATCHING,
        arrayOf(EpisodeColumns.ID, EpisodeColumns.EXPIRES_AT),
        null,
        null,
        null
      )

      val currentTime = System.currentTimeMillis()
      var expires: Long = 0
      if (watching!!.moveToFirst()) {
        expires = watching.getLong(EpisodeColumns.EXPIRES_AT)
      }

      val episode = context.contentResolver.query(
        Episodes.withId(episodeId),
        arrayOf(
          EpisodeColumns.SHOW_ID,
          EpisodeColumns.TITLE,
          EpisodeColumns.SEASON,
          EpisodeColumns.EPISODE,
          EpisodeColumns.WATCHED
        )
      )
      episode.moveToFirst()
      val showId = episode.getLong(EpisodeColumns.SHOW_ID)
      val season = episode.getInt(EpisodeColumns.SEASON)
      val number = episode.getInt(EpisodeColumns.EPISODE)
      val watched = episode.getBoolean(EpisodeColumns.WATCHED)
      val title = DataHelper.getEpisodeTitle(context, episode, season, number, watched)
      episode.close()

      val show = context.contentResolver.query(
        Shows.withId(showId),
        arrayOf(ShowColumns.RUNTIME),
        null,
        null,
        null
      )
      show!!.moveToFirst()
      val runtime = show.getInt(ShowColumns.RUNTIME)
      val watchSlop = (runtime.toFloat() * DateUtils.MINUTE_IN_MILLIS.toFloat() * 0.8f).toLong()
      show.close()

      if (watching.count == 0 || expires - watchSlop < currentTime && expires > 0) {
        if (checkIn.episode(episodeId, message, facebook, twitter, tumblr)) {
          episodeHelper.checkIn(episodeId)
        }
      } else {
        ErrorEvent.post(context.getString(R.string.checkin_error_watching, title))
      }

      watching.close()

      workManager.enqueueNow(SyncWatchingWorker::class.java)
    }
  }

  fun cancelCheckin() {
    scope.launch {
      var episode: Cursor? = null
      try {
        episode = context.contentResolver.query(
          Episodes.EPISODE_WATCHING,
          arrayOf(EpisodeColumns.ID, EpisodeColumns.STARTED_AT, EpisodeColumns.EXPIRES_AT)
        )

        if (episode.moveToFirst()) {
          val id = episode.getLong(EpisodeColumns.ID)
          val startedAt = episode.getLong(EpisodeColumns.STARTED_AT)
          val expiresAt = episode.getLong(EpisodeColumns.EXPIRES_AT)

          val values = ContentValues()
          values.put(EpisodeColumns.CHECKED_IN, false)
          context.contentResolver.update(Episodes.EPISODE_WATCHING, values, null, null)

          try {
            val call = checkinService.deleteCheckin()
            val response = call.execute()
            if (response.isSuccessful) {
              return@launch
            }
          } catch (e: IOException) {
            Timber.d(e)
          }

          ErrorEvent.post(context.getString(R.string.checkin_cancel_error))

          values.clear()
          values.put(EpisodeColumns.CHECKED_IN, true)
          values.put(EpisodeColumns.STARTED_AT, startedAt)
          values.put(EpisodeColumns.EXPIRES_AT, expiresAt)
          context.contentResolver.update(Episodes.withId(id), values, null, null)

          workManager.enqueueNow(SyncWatchingWorker::class.java)
        }
      } finally {
        episode?.close()
      }
    }
  }

  /**
   * Add episodes to user library collection.
   *
   * @param episodeId The database id of the episode.
   */
  fun setIsInCollection(episodeId: Long, inCollection: Boolean) {
    scope.launch {
      var collectedAt: String? = null
      var collectedAtMillis = 0L
      if (inCollection) {
        collectedAt = TimeUtils.getIsoTime()
        collectedAtMillis = TimeUtils.getMillis(collectedAt)
      }

      episodeHelper.setInCollection(episodeId, inCollection, collectedAtMillis)

      if (TraktLinkSettings.isLinked(context)) {
        val c = context.contentResolver.query(
          Episodes.withId(episodeId),
          arrayOf(EpisodeColumns.SHOW_ID, EpisodeColumns.SEASON, EpisodeColumns.EPISODE)
        )
        c.moveToFirst()
        val showId = c.getLong(EpisodeColumns.SHOW_ID)
        val traktId = showHelper.getTraktId(showId)
        val season = c.getInt(EpisodeColumns.SEASON)
        val number = c.getInt(EpisodeColumns.EPISODE)
        c.close()

        queue(CollectEpisode(traktId, season, number, inCollection, collectedAt))
      }
    }
  }

  fun setIsInWatchlist(episodeId: Long, inWatchlist: Boolean) {
    scope.launch {
      var listedAt: String? = null
      var listeddAtMillis = 0L
      if (inWatchlist) {
        listedAt = TimeUtils.getIsoTime()
        listeddAtMillis = TimeUtils.getMillis(listedAt)
      }

      episodeHelper.setIsInWatchlist(episodeId, inWatchlist, listeddAtMillis)

      if (TraktLinkSettings.isLinked(context)) {
        val c = context.contentResolver.query(
          Episodes.withId(episodeId),
          arrayOf(EpisodeColumns.SHOW_ID, EpisodeColumns.SEASON, EpisodeColumns.EPISODE)
        )
        c.moveToFirst()
        val showId = c.getLong(EpisodeColumns.SHOW_ID)
        val traktId = showHelper.getTraktId(showId)
        val season = c.getInt(EpisodeColumns.SEASON)
        val number = c.getInt(EpisodeColumns.EPISODE)
        c.close()

        queue(WatchlistEpisode(traktId, season, number, inWatchlist, listedAt))
      }
    }
  }

  /**
   * Rate an episode on trakt. Depending on the user settings, this will also send out social
   * updates to facebook,
   * twitter, and tumblr.
   *
   * @param episodeId The database id of the episode.
   * @param rating A rating betweeo 1 and 10. Use 0 to undo rating.
   */
  fun rate(episodeId: Long, rating: Int) {
    scope.launch {
      val ratedAt = TimeUtils.getIsoTime()
      val ratedAtMillis = TimeUtils.getMillis(ratedAt)

      val showId = episodeHelper.getShowId(episodeId)
      val showTraktId = showHelper.getTraktId(showId)
      val c = context.contentResolver.query(
        Episodes.withId(episodeId),
        arrayOf(EpisodeColumns.EPISODE, EpisodeColumns.SEASON)
      )

      if (c.moveToFirst()) {
        val episode = c.getInt(EpisodeColumns.EPISODE)
        val season = c.getInt(EpisodeColumns.SEASON)

        val values = ContentValues()
        values.put(EpisodeColumns.USER_RATING, rating)
        values.put(EpisodeColumns.RATED_AT, ratedAtMillis)
        context.contentResolver.update(Episodes.withId(episodeId), values, null, null)

        if (TraktLinkSettings.isLinked(context)) {
          queue(RateEpisode(showTraktId, season, episode, rating, ratedAt))
        }
      }
      c.close()
    }
  }
}
